esphome/components/tem3200/tem3200.cpp

```diff
@@ -0,0 +1,154 @@
+#include "tem3200.h"
+#include "esphome/core/log.h"
+#include "esphome/core/helpers.h"
+#include "esphome/core/hal.h"
+
+namespace esphome {
+namespace tem3200 {
+
+static const char *const TAG = "tem3200";
+
+enum ErrorCode {
+  NONE = 0,
+  RESERVED = 1,
+  STALE = 2,
+  FAULT = 3,
+};
+
+void TEM3200Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up TEM3200...");
+
+  uint8_t status(NONE);
+  uint16_t temperature_raw(0);
+  uint16_t pressure_raw(0);
+
+  // startup device delay
+  delay(10);
+
+  i2c::ErrorCode err = this->read_(status, temperature_raw, pressure_raw);
+  if (err != i2c::ERROR_OK) {
+    ESP_LOGCONFIG(TAG, "    I2C Communication Failed...");
+    this->mark_failed();
+    return;
+  }
+
+  switch (status) {
+    case RESERVED:
+      ESP_LOGE(TAG, "Invalid RESERVED Device Status");
+      this->mark_failed();
+      return;
+    case FAULT:
+      ESP_LOGE(TAG, "FAULT condition in the SSC or sensing element");
+      this->mark_failed();
+      return;
+    case STALE:
+      ESP_LOGE(TAG, "STALE Data. Data has been fetched since last measurement cycle");
+      break;
+  }
+  ESP_LOGCONFIG(TAG, "    Success...");
+}
+
+void TEM3200Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "TEM3200:");
+  LOG_I2C_DEVICE(this);
+  LOG_UPDATE_INTERVAL(this);
+  LOG_SENSOR("  ", "Temperature", this->temperature_sensor_);
+  LOG_SENSOR("  ", "Pressure", this->pressure_sensor_);
+}
+
+float TEM3200Component::get_setup_priority() const { return setup_priority::DATA; }
+
+i2c::ErrorCode TEM3200Component::read_(uint8_t &status, uint16_t &temperature_raw, uint16_t &pressure_raw) {
+  uint8_t response[4] = {0x00, 0x00, 0x00, 0x00};
+
+  // initiate data read
+  i2c::ErrorCode err = this->read(response, 4);
+  if (err != i2c::ERROR_OK) {
+    return err;
+  }
+
+  // extract top 2 bits of first byte for status
+  status = (ErrorCode) (response[0] & 0xc0) >> 6;
+  if (status == RESERVED || status == FAULT) {
+    return err;
+  }
+
+  // if data is stale; reread
+  if (status == STALE) {
+    // wait for measurement 2ms (as illustrated in tem3200 sample code)
+    delay(2);
+
+    err = this->read(response, 4);
+    if (err != i2c::ERROR_OK) {
+      return err;
+    }
+  }
+
+  // extract top 2 bits of first byte for status
+  status = (ErrorCode) (response[0] & 0xc0) >> 6;
+  if (status == RESERVED || status == FAULT) {
+    return err;
+  }
+
+  // extract top 6 bits of first byte and all bits of second byte for pressure
+  pressure_raw = (((response[0] & 0x3f)) << 8 | response[1]);
+
+  // extract all bytes of 3rd byte and top 3 bits of fourth byte for temperature
+  temperature_raw = ((response[2] << 3) | (response[3] & 0xe0) >> 5);
+
+  return err;
+}
+
+float TEM3200Component::convert_temperature_(uint16_t temperature_raw) {
```

Comment:
oh, I thought you were returning the raw data and letting the user convert like you do with PSI..  huh, I think this deserves a bit more comment in the docs ("driver converts the returned data as best it can to deg C based on docs but should not be relied on for accuracy" or something like that?  (I don't think I love the driver converting C and not doing anything with PSI)

Reply:
I did fixup the docs for clarity. It clarifies that the temperature is converted to C, but the pressure is left as an exersize for the developer.

From my reading of the docs, the temp conversion is reasonably well known, but the pressure isn't, and it can vary depending on the pressure range of the device.

What I did seemed like the best compromise I could come up with.

Reply:
Correct me if I am wrong:  generally the sensors have a max pressure rating which you could take as an argument to the driver and get fairly accurate PSI output without calibration (probably about as accurate as the temperature?)  You could then also put in the instructions how to calibrate for better accuracy? (Just an idea)

Reply:
That is possible.  I had major issues at the low temperature range getting reasonable values. But in the middle they were spot on.

I think I see two reasonable options.

1) Add a new option (max pressure) which allows the device to compute the actual pressure. Then both pressure and temperature will be yielded in actual units. If you want to work with the raw pressure units, you can leave that option unset, and it will return the raw data to you.

2) Add a new option (max pressure) which allows the device to compute the actual pressure. Then both pressure and temperature will yield actual units. Also add two more options raw_pressure, and raw_temperature which you can use to get the raw values directly.

Both options seem a bit more complicated than most other, initial implementations I've poked at. But I can certainly do it.